### PR TITLE
ROCm MIOpen NHWC Convolution support

### PR DESCRIPTION
**Pull request by amathews-amd**
- Added 2D-Convolution NHWC support 
  - on ROCm 4.3, with `PYTORCH_MIOPEN_SUGGEST_NHWC=1` flag
  - May need to force MIOpen to search for solutions ( see examples below for flags )
 
**PYTORCH_MIOPEN_SUGGEST_NHWC Environment Flag** 
MIOpen does not officially support NHWC yet, although convolution support has been added to tip-of-tree of MIOpen. This flag is intended to be a short-lived flag to explicitly turn on NHWC support until ROCm officially supports NHWC and performance is verified. 

**Examples**
1. Example usage 1 : Run test on ROCm4.3
`PYTORCH_TEST_WITH_ROCM=1 PYTORCH_MIOPEN_SUGGEST_NHWC=1 MIOPEN_FIND_ENFORCE=4 MIOPEN_DEBUG_CONV_GEMM=0 MIOPEN_FIND_MODE=1 pytest test_nn.py -v -k "test_conv_cudnn_nhwc" `
2. Example usage 2: Run the following with `PYTORCH_MIOPEN_SUGGEST_NHWC=1` on ROCm4.3. 
```
#!/usr/bin/env python3
import torch
model = torch.nn.Conv2d(8, 4, 3).cuda().half()
model = model.to(memory_format=torch.channels_last)
input = torch.randint(1, 10, (2, 8, 4, 4), dtype=torch.float32, requires_grad=True)
input = input.to(device="cuda", memory_format=torch.channels_last, dtype=torch.float16)

# should print True for is_contiguous(channels_last), and strides must match NHWC format
print(input.is_contiguous(memory_format=torch.channels_last), input.shape, input.stride() )

out = model(input)

# should print True for is_contiguous(channels_last), and strides must match NHWC format
print("Contiguous channel last :", out.is_contiguous(memory_format=torch.channels_last), " out shape :",  out.shape, "out stride :", out.stride() ) 
```

See https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html for more examples. 

cc @jeffdaily @sunway513 @jithunnair-amd @ROCmSupport

### CODE REVIEW

File: aten/src/ATen/native/ConvUtils.h [hunk 1]
```diff
@@ -106,4 +107,31 @@ static inline bool cudnn_conv_use_channels_last(const at::Tensor& input, const a
...
+      weight.scalar_type() == at::kDouble) {
+    return false;
+  }
+
+  auto input_memory_format = input.suggest_memory_format();
+  auto weight_memory_format = weight.suggest_memory_format();
+
+  bool can_use_miopen_channels_last_2d = false;
+#if defined(USE_ROCM) && (ROCM_VERSION >= 40300)
+  can_use_miopen_channels_last_2d = PYTORCH_MIOPEN_SUGGEST_NHWC &&  *PYTORCH_MIOPEN_SUGGEST_NHWC && (
```
Comment (jeffdaily) [inline aten/src/ATen/native/ConvUtils.h:126]: What happens when the optional<bool> is a nullopt because the user set the env var to something other than 0 or 1?  Do we care about that use case and failing gracefully?

Aside: The c10::utils::check_env appears to have been added recently (#59052), but is currently not used anywhere else.  ATen seems to use getenv directly.  Does upstream have a preference how to parse env vars?

Comment (amathews-amd) [reply aten/src/ATen/native/ConvUtils.h:126]: You can see that I am testing both `PYTORCH_MIOPEN_SUGGEST_NHWC` and `*PYTORCH_MIOPEN_SUGGEST_NHWC`. The first one is the nullopt check. The second is the true/false check. About handling other arguments like 'True/False', thats upto check_env. 

Yes, c10::utils::check_env is new, and not used. I figured I will use it since the API is supported, and it might be a direction where the PyTorch devs want to go.

File: torch/testing/_internal/common_device_type.py [hunk 2]
```diff
@@ -1167,6 +1168,32 @@ def skipCUDAIfRocm(fn):
...
+# Skips a test on CUDA if ROCm is unavailable or its version is lower than requested.
+def skipCUDAIfRocmVersionLessThan(version=None):
+
+    def dec_fn(fn):
+        @wraps(fn)
+        def wrap_fn(self, *args, **kwargs):
+            if self.device_type == 'cuda':
+                if not TEST_WITH_ROCM:
+                    reason = "ROCm not available"
+                    raise unittest.SkipTest(reason)
```
Comment (jeffdaily) [inline torch/testing/_internal/common_device_type.py:1180]: I think this will cause CUDA to skip any test using this decorator because CUDA never sets TEST_WITH_ROCM.  I think the logic you were going for was

```
if self.device_type == 'cuda' and torch.version.hip is not None
    # rocm version parsing etc
```

**Review by jeffdaily — COMMENTED** [review]
I'm confused why we need both `skipIfRocmVersionLessThan` and `skipCUDAIfRocmVersionLessThan`.

File: aten/src/ATen/native/ConvUtils.h (hunk 1 above)
Comment (jeffdaily) [reply aten/src/ATen/native/ConvUtils.h:126]: sorry, I missed the nullopt check.  looking forward to hearing from upstream reviewer how they'd like to handle env vars.

**Comment by amathews-amd** [comment]
> I'm confused why we need both `skipIfRocmVersionLessThan` and `skipCUDAIfRocmVersionLessThan`.

'device_type' is defined for a class, TestNN, but it exists for TestNNDeviceType (possibly due to `instantiate_device_type_tests(TestNNDeviceType, globals())` line in test_nn.py ). So, we can't just use `skipCUDAIfRocmVersionLessThan`. 

Can we just use `skipIfRocmVersionLessThan` ? Possibly, since the applicable tests are wrapped in `@onlyCUDA` anyway. However, this is against the design of the skipping decorator and even skip on CPU (if support for NHWC is added on CPU in future). So, I am making a conservative choice here.

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 3]
```diff
@@ -524,7 +524,20 @@ void miopen_convolution_add_bias_(CheckedFrom c, const TensorArg& output, const
...
+
+  auto memory_format = output->suggest_memory_format();
+
+  std::vector<int64_t> shape( output->dim(), 1);
+  shape[output_channels_dim] = -1;
+  at::Tensor bias_contig =  bias->reshape(shape).contiguous(memory_format);
+  // Make sure that NC11 strides follow formula
+  bias_contig.resize_(bias_contig.sizes(), memory_format );
+
+  // TODO: Workaround since MIOpen does not support NHWC bias
```
Comment (amathews-amd) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp:536]: Note: Internal tickets opened against frameworks pytorch to fix this later: 
https://ontrack-internal.amd.com/browse/SWDEV-301466
See ticket for linked MIOpen ticket for NHWC bias support.

File: torch/testing/_internal/common_device_type.py (hunk 2 above)
Comment (amathews-amd) [reply torch/testing/_internal/common_device_type.py:1180]: I am not sure how the ROCm CI sets `PYTORCH_TEST_WITH_ROCM=1`, but `TEST_WITH_ROCM` is just the internal representation of that flag (read in `common_utils.py`). I got the idea to use this flag from the `skipCUDAIfNotRocm()` function just a few lines above. 

When I am testing, I am manually setting `PYTORCH_TEST_WITH_ROCM`, as in the description of this PR.

File: aten/src/ATen/miopen/Descriptors.cpp [hunk 4]
```diff
@@ -109,9 +109,25 @@ void FilterDescriptor::set(const at::Tensor &t, int64_t pad) {
...
+  else {
+      // ChannelsLast
+       for (int i = pad; i >= dim; --i ) {
+          stride[i] = 1;
+      }
+      for (int i = dim-1 ; i >=0; --i ) {
+          // Pass-through
+          stride[i] = t.stride(i);
+      }
+  }
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.cpp:129]: I'm confused by this. Why don't you just read out the strides from the tensor, instead of recomputing them here? (Does miopen require some specific canonical form of strides?)

File: aten/src/ATen/miopen/Descriptors.cpp [hunk 5]
```diff
@@ -90,17 +90,17 @@ std::ostream& operator<<(std::ostream & out, const TensorDescriptor& d) {
 
 void TensorDescriptor::print() { std::cout << *this; }
 
-void FilterDescriptor::set(const at::Tensor &t, int64_t pad) {
+void FilterDescriptor::set(const at::Tensor &t, const at::MemoryFormat memory_format, int64_t pad) {
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.cpp:93]: I actually think passing memory format here explicitly is kind of suspect, but it is symmetric with cuDNN so I'll let it slide.

File: aten/src/ATen/miopen/Descriptors.cpp [hunk 6]
```diff
@@ -109,9 +109,25 @@ void FilterDescriptor::set(const at::Tensor &t, int64_t pad) {
   for (int i = dim; i < pad; ++i) {
     size[i] = (int) 1;
   }
-  for (int i = dim - 1; i >=0; --i) {
-    stride[i] = (i == dim - 1) ? 1 : stride[i+1] * size[i+1];
+  if( memory_format != at::MemoryFormat::ChannelsLast ) {
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.cpp:112]: Don't do this. Do a switch on memory format and explicitly error if it is an unexpected error; this will make the code robust if a new memory format gets added.

File: aten/src/ATen/native/ConvUtils.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/ConvUtils.h:126]: Can you explain why this feature has to be gated by an environment variable. For alternative layout support, ordinarily you would simply just do the correct layout algorithm depending on what the layout of the weights are. To compare, cuDNN supports alternative layouts without needing an envvar to modulate.

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:536]: link the ticket in the code

File: aten/src/ATen/miopen/Descriptors.cpp (hunk 4 above)
Comment (amathews-amd) [reply aten/src/ATen/miopen/Descriptors.cpp:129]: For the Pad dimensions, it needs to be set to 1. So, this addition is needed. 
In the ChannelsLast code, the new code is essentially doing what you propose, that is, reading out strides from the tensor. 
In the NCHW (and other code), on line 117, I only kept it to recompute because the original code was recomputing. I don't think we need to recompute. If you are also on same opinion, I can change both code to just pass-through

File: aten/src/ATen/miopen/Descriptors.cpp (hunk 5 above)
Comment (amathews-amd) [reply aten/src/ATen/miopen/Descriptors.cpp:93]: Yes, I referred to the cuDNN implementation and tried to match it as much as possible.

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 7]
```diff
@@ -593,10 +607,19 @@ Tensor miopen_convolution_forward(
   checkAllSameType(c, {input, weight});
   checkAllSameGPU(c, {input, weight});
 
-  auto output_t = at::empty(
+  auto memory_format = at::MemoryFormat::Contiguous;
+  if (miopen_conv_use_channels_last(*input, *weight)) {
+    memory_format = (weight->ndimension() == 5) ? /*at::MemoryFormat::ChannelsLast3d*/at::MemoryFormat::Contiguous : at::MemoryFormat::ChannelsLast;
+  }
+
+  auto output_t = at::native::empty_cuda(
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:615]: It looks like this was cargo culted from ConvShared.cpp

File: aten/src/ATen/native/ConvUtils.h (hunk 1 above)
Comment (amathews-amd) [reply aten/src/ATen/native/ConvUtils.h:126]: The reason it has to be gated is because MIOpen does not officially support NHWC yet, although convolution support has been added to tip-of-tree of MIOpen. The plan is to remove the environment variable check to a ROCm version check once support is officially added. 
( Also, the MIOpen teams do need this support in PyTorch during this time, since we are testing performance of the newly added NHWC code on application loads. )

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 8]
```diff
@@ -1141,6 +1245,24 @@ Tensor miopen_convolution_backward_bias(
...
+  std::vector<int64_t> discard_dims;
+  for( int i = 0; i < grad_output_t.dim(); i++ ) {
+      if(i != output_channels_dim ) {
+          discard_dims.push_back(i);
+      }
+  }
+
+  Tensor outputBias = at::squeeze( at::sum(grad_output_t, discard_dims, true) );
+  if( outputBias.dim() == 0 ) {
+      // always return a tensor of shape [_]
```
Comment (ezyang) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp:1258]: How come?

File: aten/src/ATen/miopen/Descriptors.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/miopen/Descriptors.cpp:129]: OK I see. You don't have to fix the original code in this PR but I would like to see it fixed (Unless there is a reason for recompute, in which case it should be commented why)

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 8 above)
Comment (amathews-amd) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:1258]: If the result is just 1 element, 1 dim, at::squeeze will make the result a scalar. There are tests in test_nn.py that check this return value against a tensor of shape, [1].

**Comment by ezyang** [comment]
This looks ok but I'm skeptical about the environment variable

File: aten/src/ATen/native/ConvUtils.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/ConvUtils.h:126]: OK, but it's pretty difficult to "accidentally" end up running NHWC when you didn't intend to (the weights have to be NHWC). Wouldn't you just rather error in that case for now?

Comment (ezyang) [reply aten/src/ATen/native/ConvUtils.h:126]: In any case, the plan on record was not clear from the PR description nor the code

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 9]
```diff
@@ -524,7 +524,20 @@ void miopen_convolution_add_bias_(CheckedFrom c, const TensorArg& output, const
...
   TensorDescriptor bdesc, odesc;
-  bdesc.set(bias->expand({1, bias->size(0)}), output->dim());
+
+  auto memory_format = output->suggest_memory_format();
+
+  std::vector<int64_t> shape( output->dim(), 1);
+  shape[output_channels_dim] = -1;
+  at::Tensor bias_contig =  bias->reshape(shape).contiguous(memory_format);
+  // Make sure that NC11 strides follow formula
+  bias_contig.resize_(bias_contig.sizes(), memory_format );
```
[the remark this answers is not in the document]
Comment (amathews-amd) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:534]: I agree that its not optimal to keep doing this. I was following the pattern in the cudnn implementation. 

The NHWC blog (https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html) describes this issue. 
`For general cases the two APIs behave the same. However in special cases for a 4D tensor with size NCHW when either: C==1 or H==1 && W==1, only to would generate a proper stride to represent channels last memory format.`

File: aten/src/ATen/miopen/Descriptors.cpp (hunk 4 above)
Comment (amathews-amd) [reply aten/src/ATen/miopen/Descriptors.cpp:119]: I decided to fix the original code; Tested locally and test_nn passes. Lets see if CI tests pass. If it does, we are good.

File: aten/src/ATen/miopen/Descriptors.cpp (hunk 6 above)
Comment (amathews-amd) [reply aten/src/ATen/miopen/Descriptors.cpp]: Updated; this is gone.

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (amathews-amd) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:616]: This `empty_cuda` function sets the strides correctly given the memory layout, which is why I used it.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Sorry about the delay. Need some warning cleanup in this PR:

```
caffe2/aten/src/ATen/native/ConvUtils.h:114:30: error: unused variable 'PYTORCH_MIOPEN_SUGGEST_NHWC' [-Werror,-Wunused-variable]
  static c10::optional<bool> PYTORCH_MIOPEN_SUGGEST_NHWC = c10::utils::check_env("PYTORCH_MIOPEN_SUGGEST_NHWC");
                             ^
caffe2/aten/src/ATen/native/ConvUtils.h:123:8: error: unused variable 'input_memory_format' [-Werror,-Wunused-variable]
  auto input_memory_format = input.suggest_memory_format();
       ^
caffe2/aten/src/ATen/native/ConvUtils.h:124:8: error: unused variable 'weight_memory_format' [-Werror,-Wunused-variable]
  auto weight_memory_format = weight.suggest_memory_format();
```

**Comment by amathews-amd** [comment]
> Sorry about the delay. Need some warning cleanup in this PR:
> 
> ```
> caffe2/aten/src/ATen/native/ConvUtils.h:114:30: error: unused variable 'PYTORCH_MIOPEN_SUGGEST_NHWC' [-Werror,-Wunused-variable]
>   static c10::optional<bool> PYTORCH_MIOPEN_SUGGEST_NHWC = c10::utils::check_env("PYTORCH_MIOPEN_SUGGEST_NHWC");
>                              ^
> caffe2/aten/src/ATen/native/ConvUtils.h:123:8: error: unused variable 'input_memory_format' [-Werror,-Wunused-variable]
>   auto input_memory_format = input.suggest_memory_format();
>        ^
> caffe2/aten/src/ATen/native/ConvUtils.h:124:8: error: unused variable 'weight_memory_format' [-Werror,-Wunused-variable]
>   auto weight_memory_format = weight.suggest_memory_format();
> ```

Fixed, and merged upstream to branch.
